File: adapter/transform.py
from .config import DEFAULT_REASONING_EFFORT, DEFAULT_THINKING_TYPE, is_deepseek_v4_model, normalize_reasoning_effort, normalize_thinking_type
from .content import content_to_chat, request_has_image
from .jsonutil import json_dumps
from .reasoning import extract_reasoning_content, lookup_reasoning


def responses_input_to_messages(request, server=None):
    messages = []
    instructions = request.get("instructions")
    if instructions:
        messages.append({"role": "system", "content": str(instructions)})

    response_input = request.get("input", "")
    if isinstance(response_input, str):
        if response_input:
            messages.append({"role": "user", "content": response_input})
        return messages

    if not isinstance(response_input, list):
        messages.append({"role": "user", "content": content_to_chat(response_input)})
        return messages

    assistant_message = None

    def ensure_assistant():
        nonlocal assistant_message
        if assistant_message is None:
            assistant_message = {"role": "assistant", "content": ""}
        return assistant_message

    def flush_assistant():
        nonlocal assistant_message
        if assistant_message is None:
            return
        if not assistant_message.get("tool_calls"):
            assistant_message.pop("tool_calls", None)
        if not assistant_message.get("reasoning_content"):
            assistant_message.pop("reasoning_content", None)
        messages.append(assistant_message)
        assistant_message = None

    def append_message_item(item):
        role = item.get("role", "user")
        if role in ("developer", "system"):
            role = "system"
        elif role not in ("user", "assistant", "tool"):
            role = "user"
        content = content_to_chat(item.get("content"))
        if role == "assistant":
            assistant = ensure_assistant()
            if content:
                assistant["content"] = content if not assistant.get("content") else assistant["content"] + "\n" + content
            reasoning_content = extract_reasoning_content(item)
            if reasoning_content:
                assistant["reasoning_content"] = reasoning_content
        else:
            flush_assistant()
            messages.append({"role": role, "content": content})

    for item in response_input:
        if not isinstance(item, dict):
            flush_assistant()
            messages.append({"role": "user", "content": content_to_chat(item)})
            continue

        item_type = item.get("type")
        if item_type == "message" or (item_type is None and ("role" in item or "content" in item)):
            append_message_item(item)
        elif item_type == "function_call":
            call_id = item.get("call_id") or item.get("id") or "call_unknown"
            assistant = ensure_assistant()
            if not assistant.get("reasoning_content"):
                assistant["reasoning_content"] = extract_reasoning_content(item) or lookup_reasoning(server, call_id)
            assistant.setdefault("tool_calls", []).append(
                {
                    "id": call_id,
                    "type": "function",
                    "function": {
                        "name": item.get("name", ""),
                        "arguments": item.get("arguments", "{}"),
                    },
                }
            )
        elif item_type == "function_call_output":
            flush_assistant()
            call_id = item.get("call_id") or item.get("id") or "call_unknown"
            messages.append({"role": "tool", "tool_call_id": call_id, "content": content_to_chat(item.get("output"))})
        elif item_type == "reasoning":
            reasoning_content = extract_reasoning_content(item)
            if reasoning_content:
                ensure_assistant()["reasoning_content"] = reasoning_content
        else:
            flush_assistant()
            messages.append({"role": "user", "content": json_dumps(item)})

    flush_assistant()
    return messages


def responses_tools_to_chat_tools(tools):
    chat_tools = []
    for tool in tools or []:
        if not isinstance(tool, dict) or tool.get("type") != "function":
            continue
        name = tool.get("name")
        if not name:
            continue
        chat_tools.append(
            {
                "type": "function",
                "function": {
                    "name": name,
                    "description": tool.get("description", ""),
                    "parameters": tool.get("parameters", {"type": "object", "properties": {}}),
                },
            }
        )
    return chat_tools


def responses_to_chat_request(request, default_model, server=None):
    chat_request = {
        "model": request.get("model") or default_model,
        "messages": responses_input_to_messages(request, server),
        "stream": bool(request.get("stream", True)),
    }
    tools = responses_tools_to_chat_tools(request.get("tools"))
    if tools:
        chat_request["tools"] = tools
        chat_request["tool_choice"] = "auto"
    if "tool_choice" in request and request["tool_choice"] in ("none", "auto", "required"):
        chat_request["tool_choice"] = request["tool_choice"]
    if "response_format" in request:
        chat_request["response_format"] = request["response_format"]
    if "stop" in request:
        chat_request["stop"] = request["stop"]
    if "stream_options" in request:
        chat_request["stream_options"] = request["stream_options"]
    if "user" in request:
        chat_request["user_id"] = request["user"]
    if "user_id" in request:
        chat_request["user_id"] = request["user_id"]
    if "frequency_penalty" in request:
        chat_request["frequency_penalty"] = request["frequency_penalty"]
    if "presence_penalty" in request:
        chat_request["presence_penalty"] = request["presence_penalty"]

    deepseek_v4 = is_deepseek_v4_model(chat_request["model"])
    if "temperature" in request and not deepseek_v4:
        chat_request["temperature"] = request["temperature"]
    if "top_p" in request and not deepseek_v4:
        chat_request["top_p"] = request["top_p"]
    if "max_output_tokens" in request:
        chat_request["max_tokens"] = request["max_output_tokens"]
    if "max_tokens" in request:
        chat_request["max_tokens"] = request["max_tokens"]

    reasoning_effort = None
    reasoning = request.get("reasoning")
    if isinstance(reasoning, dict):
        reasoning_effort = reasoning.get("effort")
        thinking_type = reasoning.get("thinking")
    else:
        thinking_type = None
    reasoning_effort = request.get("reasoning_effort") or reasoning_effort
    thinking_type = request.get("thinking") or thinking_type
    if deepseek_v4:
        normalized_thinking = normalize_thinking_type(
            thinking_type,
            default=getattr(server, "default_thinking_type", DEFAULT_THINKING_TYPE),
        )
        chat_request["thinking"] = {"type": normalized_thinking}
    if not reasoning_effort and deepseek_v4:
        reasoning_effort = getattr(server, "default_reasoning_effort", DEFAULT_REASONING_EFFORT)
    if reasoning_effort and (deepseek_v4 or request.get("reasoning_effort")):
        normalized_effort = normalize_reasoning_effort(reasoning_effort, default=None)
        if normalized_effort:
            chat_request["reasoning_effort"] = normalized_effort
    return chat_request


def select_model(request, path, default_model, compact_model, vision_model):
    if request_has_image(request):
        return vision_model
    if path.endswith("/compact"):
        return compact_model
    return request.get("model") or default_model

File: internal/transform/transform.go
package transform

import (
	"github.com/Hiyajomaho-num9/opencode-go-codex/internal/config"
	"github.com/Hiyajomaho-num9/opencode-go-codex/internal/content"
	"github.com/Hiyajomaho-num9/opencode-go-codex/internal/jsonutil"
	"github.com/Hiyajomaho-num9/opencode-go-codex/internal/reasoning"
)

type Defaults struct {
	DefaultReasoningEffort string
	DefaultThinkingType    string
	ReasoningStore         *reasoning.Store
}

func ResponsesInputToMessages(request map[string]any, defaults Defaults) []map[string]any {
	messages := []map[string]any{}
	if instructions, ok := request["instructions"].(string); ok && instructions != "" {
		messages = append(messages, map[string]any{"role": "system", "content": instructions})
	}

	responseInput, ok := request["input"]
	if !ok {
		return messages
	}
	if s, ok := responseInput.(string); ok {
		if s != "" {
			messages = append(messages, map[string]any{"role": "user", "content": s})
		}
		return messages
	}
	items, ok := responseInput.([]any)
	if !ok {
		messages = append(messages, map[string]any{"role": "user", "content": content.ToChat(responseInput)})
		return messages
	}

	var assistant map[string]any
	ensureAssistant := func() map[string]any {
		if assistant == nil {
			assistant = map[string]any{"role": "assistant", "content": ""}
		}
		return assistant
	}
	flushAssistant := func() {
		if assistant == nil {
			return
		}
		if calls, ok := assistant["tool_calls"].([]any); !ok || len(calls) == 0 {
			delete(assistant, "tool_calls")
		}
		if rc, ok := assistant["reasoning_content"].(string); !ok || rc == "" {
			delete(assistant, "reasoning_content")
		}
		messages = append(messages, assistant)
		assistant = nil
	}

	appendMessageItem := func(item map[string]any) {
		role, _ := item["role"].(string)
		if role == "" {
			role = "user"
		}
		if role == "developer" || role == "system" {
			role = "system"
		} else if role != "user" && role != "assistant" && role != "tool" {
			role = "user"
		}
		chatContent := content.ToChat(item["content"])
		if role == "assistant" {
			a := ensureAssistant()
			if s, ok := chatContent.(string); ok && s != "" {
				if existing, _ := a["content"].(string); existing != "" {
					a["content"] = existing + "\n" + s
				} else {
					a["content"] = s
				}
			} else if chatContent != nil {
				a["content"] = chatContent
			}
			itemType, _ := item["type"].(string)
			if itemType == "reasoning" || item["encrypted_content"] != nil || item["reasoning_content"] != nil {
				if rc := reasoning.Extract(item); rc != "" {
					a["reasoning_content"] = rc
				}
			}
			return
		}
		flushAssistant()
		messages = append(messages, map[string]any{"role": role, "content": chatContent})
	}

	for _, raw := range items {
		item, ok := raw.(map[string]any)
		if !ok {
			flushAssistant()
			messages = append(messages, map[string]any{"role": "user", "content": content.ToChat(raw)})
			continue
		}
		itemType, _ := item["type"].(string)
		switch {
		case itemType == "message" || (itemType == "" && (item["role"] != nil || item["content"] != nil)):
			appendMessageItem(item)
		case itemType == "function_call":
			callID := firstString(item["call_id"], item["id"], "call_unknown")
			a := ensureAssistant()
			if _, ok := a["reasoning_content"]; !ok {
				if rc := reasoning.Extract(item); rc != "" {
					a["reasoning_content"] = rc
				} else if defaults.ReasoningStore != nil {
					if rc := defaults.ReasoningStore.Lookup(callID); rc != "" {
						a["reasoning_content"] = rc
					}
				}
			}
			calls, _ := a["tool_calls"].([]any)
			calls = append(calls, map[string]any{
				"id":   callID,
				"type": "function",
				"function": map[string]any{
					"name":      firstString(item["name"], "", ""),
					"arguments": firstString(item["arguments"], "{}", "{}"),
				},
			})
			a["tool_calls"] = calls
		case itemType == "function_call_output":
			flushAssistant()
			callID := firstString(item["call_id"], item["id"], "call_unknown")
			messages = append(messages, map[string]any{"role": "tool", "tool_call_id": callID, "content": content.ToChat(item["output"])})
		case itemType == "reasoning":
			if rc := reasoning.Extract(item); rc != "" {
				ensureAssistant()["reasoning_content"] = rc
			}
		default:
			flushAssistant()
			messages = append(messages, map[string]any{"role": "user", "content": jsonutil.MustMarshalString(item)})
		}
	}
	flushAssistant()
	return messages
}

func ResponsesToolsToChatTools(tools any) []any {
	items, ok := tools.([]any)
	if !ok {
		return nil
	}
	out := []any{}
	for _, raw := range items {
		tool, ok := raw.(map[string]any)
		if !ok || tool["type"] != "function" {
			continue
		}
		name, _ := tool["name"].(string)
		if name == "" {
			continue
		}
		params := tool["parameters"]
		if params == nil {
			params = map[string]any{"type": "object", "properties": map[string]any{}}
		}
		out = append(out, map[string]any{
			"type": "function",
			"function": map[string]any{
				"name":        name,
				"description": firstString(tool["description"], "", ""),
				"parameters":  params,
			},
		})
	}
	return out
}

func ResponsesToChatRequest(request map[string]any, model string, defaults Defaults) map[string]any {
	if model == "" {
		model = firstString(request["model"], "", "")
	}
	chat := map[string]any{
		"model":    model,
		"messages": ResponsesInputToMessages(request, defaults),
		"stream":   boolValue(request["stream"], true),
	}
	if tools := ResponsesToolsToChatTools(request["tools"]); len(tools) > 0 {
		chat["tools"] = tools
		chat["tool_choice"] = "auto"
	}
	if tc, ok := request["tool_choice"].(string); ok && (tc == "none" || tc == "auto" || tc == "required") {
		chat["tool_choice"] = tc
	}
	copyFields(chat, request, map[string]string{
		"response_format":   "response_format",
		"stop":              "stop",
		"stream_options":    "stream_options",
		"frequency_penalty": "frequency_penalty",
		"presence_penalty":  "presence_penalty",
	})
	if user, ok := request["user"]; ok {
		chat["user_id"] = user
	}
	if user, ok := request["user_id"]; ok {
		chat["user_id"] = user
	}
	deepseekV4 := config.IsDeepSeekV4Model(model)
	if !deepseekV4 {
		copyFields(chat, request, map[string]string{"temperature": "temperature", "top_p": "top_p"})
	}
	if max, ok := request["max_output_tokens"]; ok {
		chat["max_tokens"] = max
	}
	if max, ok := request["max_tokens"]; ok {
		chat["max_tokens"] = max
	}
	reasoningEffort := ""
	thinkingType := ""
	if r, ok := request["reasoning"].(map[string]any); ok {
		reasoningEffort = firstString(r["effort"], "", "")
		thinkingType = firstString(r["thinking"], "", "")
	}
	if s := firstString(request["reasoning_effort"], "", ""); s != "" {
		reasoningEffort = s
	}
	if s := firstString(request["thinking"], "", ""); s != "" {
		thinkingType = s
	}
	if deepseekV4 {
		if thinkingType == "" {
			thinkingType = defaults.DefaultThinkingType
		}
		chat["thinking"] = map[string]any{"type": config.NormalizeThinkingType(thinkingType)}
		if reasoningEffort == "" {
			reasoningEffort = defaults.DefaultReasoningEffort
		}
	}
	if reasoningEffort != "" && (deepseekV4 || request["reasoning_effort"] != nil) {
		chat["reasoning_effort"] = config.NormalizeReasoningEffort(reasoningEffort)
	}
	return chat
}

func SelectModel(request map[string]any, path, defaultModel, compactModel, visionModel string) string {
	if content.RequestHasImage(request) {
		return visionModel
	}
	if len(path) >= len("/compact") && path[len(path)-len("/compact"):] == "/compact" {
		return compactModel
	}
	return firstString(request["model"], defaultModel, defaultModel)
}

func firstString(a any, b any, fallback string) string {
	for _, v := range []any{a, b} {
		if s, ok := v.(string); ok && s != "" {
			return s
		}
	}
	return fallback
}

func boolValue(value any, fallback bool) bool {
	if value == nil {
		return fallback
	}
	if b, ok := value.(bool); ok {
		return b
	}
	return fallback
}

func copyFields(dst, src map[string]any, fields map[string]string) {
	for from, to := range fields {
		if value, ok := src[from]; ok {
			dst[to] = value
		}
	}
}
